pkg/keycloak/realm/realmHandler_test.go

```diff
@@ -0,0 +1,155 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestRealmHandler(t *testing.T) {
+	kcf := &kcfake.KeycloakClientFactory{}
+	kcfsdk := &kcfake.SDKCruder{}
+	testee := NewRealmHandler(kcf, kcfsdk, kcfake.PhaseHandler{})
+	now := metav1.NewTime(time.Now())
+	cases := []struct {
+		Name          string
+		Context       context.Context
+		Object        *v1alpha1.KeycloakRealm
+		Deleted       bool
+		ShouldError   bool
+		ErrorSubstr   string
+		ExpectedState v1alpha1.StatusPhase
+	}{
+		{
+			Name:    "No error when deleted is true",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.NoPhase,
+				},
+			},
+			Deleted:       true,
+			ShouldError:   false,
+			ErrorSubstr:   "",
+			ExpectedState: v1alpha1.NoPhase,
+		},
+		{
+			Name:    "No error when deletion timestamp is set",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.NoPhase,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					DeletionTimestamp: &now,
+				},
+			},
+			ExpectedState: v1alpha1.PhaseDeprovisioned,
+		},
+		{
+			Name:    "No error when deletion timestamp is set",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.NoPhase,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					DeletionTimestamp: &now,
+				},
+			},
+			ExpectedState: v1alpha1.PhaseDeprovisioned,
+		},
+		{
+			Name:    "No error when deletion timestamp is set and phase deprovisioned",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.PhaseDeprovisioned,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					DeletionTimestamp: &now,
+				},
+			},
+			ExpectedState: v1alpha1.PhaseComplete,
+		},
+		{
+			Name:    "No error when no phase set",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.NoPhase,
+				},
+			},
+			ExpectedState: v1alpha1.NoPhase,
+		},
+		{
+			Name:    "No error when accepted phase set",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.PhaseAccepted,
+				},
+			},
+			ExpectedState: v1alpha1.PhaseAccepted,
+		},
+		{
+			Name:    "No error when provision phase set",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.PhaseProvision,
+				},
+			},
+			ExpectedState: v1alpha1.PhaseProvision,
+		},
+		{
+			Name:    "No error when provisioned phase set",
+			Context: context.TODO(),
+			Object: &v1alpha1.KeycloakRealm{
+				Status: v1alpha1.KeycloakRealmStatus{
+					Phase: v1alpha1.PhaseProvisioned,
+				},
+			},
+			ExpectedState: v1alpha1.PhaseProvisioned,
+		},
+	}
```

Comment:
do we want to cover a state when there is an error that there should be no phase change ?

Reply:
added